### PR TITLE
misc: add parts for flightrec reader in EE

### PR DESCRIPTION
**Pull request by nshy**
This includes prbuf reader from file API.

Part of https://github.com/tarantool/tarantool-ee/issues/319

EE part: https://github.com/tarantool/tarantool-ee/pull/419

### CODE REVIEW

File: src/lib/core/exception.cc [hunk 1]
```diff
@@ -431,6 +446,20 @@ BuildRaftError(const char *file, unsigned line, const char *format, ...)
...
+struct error *
+BuildFileFormatError(const char *file, unsigned line, const char *fileformat,
+		     const char *format, ...)
+{
+	BuildAlloc(FileFormatError);
+	FileFormatError *e = new(p) FileFormatError(file, line, fileformat,
+						    "");
+	va_list ap;
+	va_start(ap, format);
+	error_vformat_msg(e, format, ap);
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lib/core/exception.cc:458]: Done.

File: src/lib/core/prbuf.c [hunk 2]
```diff
@@ -369,3 +376,166 @@ prbuf_iterator_next(struct prbuf_iterator *iter, struct prbuf_entry *result)
...
+	 * Read ahead does not make sense when reading prbuf header as
+	 * regularly first record is not at the beginning of the buffer.
+	 */
+	ssize_t rc = fio_pread(fd, &header, prbuf_header_size, offset);
+	if (rc == -1) {
+		diag_set(SystemError, "read failed");
+		return -1;
+	}
+	if ((size_t)rc < prbuf_header_size) {
+		diag_set(FileFormatError, "prbuf", "can't read buffer header");
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lib/core/prbuf.c:427]: Done.

File: src/box/lua/load_cfg.lua [hunk 3]
```diff
@@ -57,16 +57,12 @@ local function ifdef_audit(value)
 end
 
 -- Flight recorder.
-local has_flightrec, flightrec = pcall(require, 'flightrec')
-
 local function ifdef_flightrec(value)
-    if has_flightrec then
+    if box.internal.flightrec_cfg then
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/lua/load_cfg.lua:61]: Done.

File: src/lib/core/say.h [hunk 4]
```diff
@@ -52,14 +52,14 @@ extern pid_t log_pid;
...
-	S_VERBOSE,
-	S_DEBUG
+	S_FATAL = 0,		/* do not use this value directly */
+	S_SYSERROR = 1,
+	S_ERROR = 2,
+	S_CRIT = 3,
+	S_WARN = 4,
+	S_INFO = 5,
+	S_VERBOSE = 6,
+	S_DEBUG = 7,
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lib/core/say.h:62]: Removed. However sometimes it useful to translate some known log level number to symbolic representation.

File: src/lib/core/say.c [hunk 5]
```diff
@@ -160,6 +160,14 @@ level_to_string(int level)
 	return level_strs[level];
 }
 
+const char*
+say_level_str(enum say_level level)
+{
+	if (level < 0 || level >= lengthof(level_strs))
+		return "unknown";
+	return level_to_string(level);
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lib/core/say.c:168]: Changed to:
```
-static const char*
-level_to_string(int level)
+const char*
+say_log_level_str(int level)
 {
-       assert(level >= S_FATAL && level <= S_DEBUG);
+       if (level < 0 || level >= (int)lengthof(level_strs))
+               return NULL;
        return level_strs[level];
 }
```
Add made some extra changes we discussed f2f.

File: src/lib/core/say.h [hunk 6]
```diff
@@ -296,6 +296,13 @@ say_set_flightrec_log_level(int new_level);
 enum say_format
 say_format_by_name(const char *format);
 
+/**
+ * Return string representation for given level. If level is not valid then
+ * return "unknown".
+ */
+const char *
+say_level_str(enum say_level level);
+
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/lib/core/say.h:305]: Done. Also add tests for new function.

File: src/box/lua/iproto.h [hunk 7]
```diff
@@ -12,6 +12,12 @@ extern "C" {
 
 struct lua_State;
 
+/**
+ * Translation table for `box.iproto.key` constants encoding and aliasing: used
+ * in `luamp_encode_with_translation` and `luamp_push_with_translation`.
+ */
+extern struct mh_strnu32_t *iproto_key_translation;
```
[the remark this answers is not in the document]
Comment (nshy) [reply src/box/lua/iproto.h:19]: Done.

**Review by locker — APPROVED** [review]

File: src/lib/core/say.c [hunk 8]
```diff
@@ -153,10 +148,13 @@ static const char *level_strs[] = {
 	[S_DEBUG] = "DEBUG",
 };
 
-static const char*
-level_to_string(int level)
+static_assert(lengthof(level_strs) == say_level_MAX);
```
Comment (locker) [inline src/lib/core/say.c:151]: This doesn't compile on OS X, because there `static_assert` must have a message.

File: src/lib/core/prbuf.c [hunk 9]
```diff
@@ -369,3 +376,166 @@ prbuf_iterator_next(struct prbuf_iterator *iter, struct prbuf_entry *result)
...
+
+int
+prbuf_reader_create(struct prbuf_reader *reader, int fd, off_t offset)
+{
+	struct prbuf_header header;
+	/*
+	 * Read ahead does not make sense when reading prbuf header as
+	 * regularly first record is not at the beginning of the buffer.
+	 */
+	ssize_t rc = fio_pread(fd, &header, prbuf_header_size, offset);
```
Comment (locker) [inline src/lib/core/prbuf.c:421]: Let's read the header lazily so that the constructor never fails. The API will look simpler then.

**Review by locker — CHANGES_REQUESTED** [review]

File: src/lib/core/say.c (hunk 8 above)
Comment (nshy) [reply src/lib/core/say.c:151]: Fixed to:
```c
static_assert(lengthof(level_strs) == say_level_MAX,
              "level_strs is not defined for one of log levels");
```
here and in the other case.

File: src/lib/core/prbuf.c (hunk 9 above)
Comment (nshy) [reply src/lib/core/prbuf.c:421]: Done.

**Pull request merged**